File: benchmarks/ts_compare/bench_streams.ts
// Stream throughput (TS reference) over the in-process MessageChannel,
// matching benchmarks/bench_streams.py: a server method returns a
// ReadableStream of 64 KiB byte chunks; the client drains it. Reports MB/s.

import { MessageChannel } from "node:worker_threads";
import { newMessagePortRpcSession, RpcTarget } from "capnweb";
import { Result } from "./harness.js";

const CHUNK = new Uint8Array(64 * 1024).fill(0xab);
const TOTAL = 64 * 1024 * 1024;

class StreamService extends RpcTarget {
  produce(): ReadableStream {
    const nChunks = TOTAL / CHUNK.length;
    let i = 0;
    return new ReadableStream({
      pull(controller) {
        if (i < nChunks) { controller.enqueue(CHUNK); i++; }
        else controller.close();
      },
    });
  }
}

async function pumpOnce(): Promise<number> {
  const { port1, port2 } = new MessageChannel();
  newMessagePortRpcSession(port2 as any, new StreamService());
  const client = newMessagePortRpcSession(port1 as any) as any;
  const t0 = process.hrtime.bigint();
  const stream: ReadableStream = await client.produce();
  const reader = stream.getReader();
  let received = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    received += (value as Uint8Array).length;
  }
  const t1 = process.hrtime.bigint();
  port1.close(); port2.close();
  if (received !== TOTAL) throw new Error(`got ${received} != ${TOTAL}`);
  return Number(t1 - t0);
}

export async function run(): Promise<Result[]> {
  await pumpOnce(); // warmup
  const samples: number[] = [];
  for (let i = 0; i < 7; i++) samples.push(await pumpOnce());
  samples.sort((a, b) => a - b);
  const median = samples[Math.floor(samples.length / 2)];
  const p95 = samples[Math.min(samples.length - 1, Math.round(0.95 * (samples.length - 1)))];
  const mbps = (TOTAL / (1024 * 1024)) / (median / 1e9);
  return [{
    name: "stream/64MiB_msgport", group: "stream.throughput",
    samples: samples.length, inner: 1, bytes: TOTAL,
    ns_median: median, ns_p95: p95,
    ns_mean: samples.reduce((a, b) => a + b, 0) / samples.length,
    ns_stddev: 0, ns_min: samples[0], ops_per_sec: 0,
    extra: { MB_per_sec: Math.round(mbps * 10) / 10, MiB_total: TOTAL / (1024 * 1024), chunk_bytes: CHUNK.length },
  }];
}

File: benchmarks/bench_streams.py
"""Stream throughput over the loopback pipe + FlowController steady state.

A server method returns an ``RpcReadableStream`` that yields fixed-size byte
chunks; the client consumes it with ``async for``. This drives the full
``["pipe"]`` + flow-controlled write/ack machinery end-to-end in one process,
so the number is the CPU-bound ceiling (no network). We report MB/s and the
FlowController's steady-state window at the end of the pump.
"""

from __future__ import annotations

import asyncio
import time
from typing import Any

from capnweb.streams import RpcReadableStream
from capnweb.types import RpcTarget

from benchmarks._harness import Result
from benchmarks._targets import PipePair


CHUNK = b"\xab" * (64 * 1024)  # 64 KiB chunks


class StreamService(RpcTarget):
    def __init__(self, total_bytes: int) -> None:
        self._total = total_bytes

    async def call(self, method: str, args: list[Any]) -> Any:
        if method == "produce":
            n_chunks = self._total // len(CHUNK)

            def gen():
                async def agen():
                    for _ in range(n_chunks):
                        yield CHUNK
                return agen()

            return RpcReadableStream(gen())
        raise Exception(f"StreamService: no method {method}")

    async def get_property(self, name: str) -> Any:
        raise Exception(f"no prop {name}")


def _observe_window() -> float | None:
    """Best-effort: read a representative steady-state window value.

    The FlowController lives inside the WritableStreamHook created per pipe;
    we can't reach it from here without touching src, so we sample the module
    default and report the tuning constants for the report instead.
    """
    from capnweb import streams

    return float(streams.INITIAL_WINDOW)


def run() -> list[Result]:
    loop = asyncio.new_event_loop()
    TOTAL = 64 * 1024 * 1024  # 64 MiB

    async def pump_once(encoding_level: str) -> float:
        pair = PipePair(StreamService(TOTAL), encoding_level=encoding_level)
        received = 0
        t0 = time.perf_counter_ns()
        stream = await pair.client.produce()
        async for chunk in stream:
            received += len(chunk)
        t1 = time.perf_counter_ns()
        await pair.stop()
        assert received == TOTAL, (received, TOTAL)
        return (t1 - t0)

    from benchmarks._harness import _stats

    def measure(name: str, encoding_level: str) -> Result:
        # Warmup (smaller): 1 full pump.
        loop.run_until_complete(pump_once(encoding_level))
        samples_ns: list[float] = []
        for _ in range(7):
            samples_ns.append(loop.run_until_complete(pump_once(encoding_level)))
        samples_ns.sort()
        median_ns = samples_ns[len(samples_ns) // 2]
        mbps = (TOTAL / (1024 * 1024)) / (median_ns / 1e9)
        s = _stats(samples_ns)
        # Note: here ns is per-PUMP (64 MiB), not per-op.
        return Result(
            name=name, group="stream.throughput",
            samples=len(samples_ns), inner=1, bytes=TOTAL,
            ns_median=s["ns_median"], ns_p95=s["ns_p95"], ns_mean=s["ns_mean"],
            ns_stddev=s["ns_stddev"], ns_min=s["ns_min"],
            ops_per_sec=0.0,
            extra={
                "MB_per_sec": round(mbps, 1),
                "MiB_total": TOTAL // (1024 * 1024),
                "chunk_bytes": len(CHUNK),
                "encoding_level": encoding_level,
                "initial_window_bytes": _observe_window(),
            },
        )

    try:
        results = [
            # Realistic WebSocket wire: every chunk is base64+JSON per protocol.
            measure("stream/64MiB_loopback", "string"),
            # P4: custom-encoding transport keeps bytes raw inside ["bytes", ...]
            # and skips the per-frame json.dumps — chunks bypass base64+JSON
            # entirely, the analog to TS's structured-clone MessagePort.
            measure("stream/64MiB_loopback_bytes", "jsonCompatibleWithBytes"),
        ]
    finally:
        loop.close()

    return results


if __name__ == "__main__":
    for r in run():
        print(f"{r.name}: {r.extra['MB_per_sec']} MB/s "
              f"(median {r.ns_median/1e6:.1f} ms/64MiB, "
              f"p95 {r.ns_p95/1e6:.1f} ms), window0={r.extra['initial_window_bytes']}")
